text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1386 @@
+that were discussed during the original Angle Brackets RFC, and holistically, we
+feel those features make a much simpler and more ergonomic component API. Taken
+alone, they are an incremental change. Their individual features aren't _that_
+much more than what we currently have in Ember today. But as a whole they
+represent the culmination of multiple years of design work and discussion by the
+Ember community, and the collective attention to detail and care of all of our
+community members.
+
+## Terminology
+
+* The **Glimmer VM** is the underlying rendering engine which is used by
+  _Ember.js_ and _Glimmer.js_.
+* **Glimmer.js** is a thin wrapper on top of the _Glimmer VM_ which exposes a
+  much simpler API compared to Ember. Historically it has been used to
+  experiment with ideas and implementations before bringing them into Ember via
+  RFC, and has been used to write applications which don't require the full
+  feature set of Ember.
+* **Glimmer components** are a newly proposed component API which draw from the
+  experimental APIs provided in Glimmer.js, and Ember.js via
+  [sparkles-component](https://github.com/rwjblue/sparkles-component).
+* **Classic components** refer to the standard component API at the time of this
+  RFC, which have been available in Ember in some form since v1.
+* **Tracked properties** refer to a new method of change tracking which is being
+  proposed in a separate RFC, parallel to this one.
+
+## Motivation
+
+`GlimmerComponent` is a simpler base component class that enables smaller class
+definitions, stronger conventions for lifecycle hooks and properties, and
+unidirectional data flow. We aim to design them to be easier understand in
+isolation, and require less knowledge of the framework to use effectively.
+
+This example shows a component written with the classic component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+{{#if (eq type 'image'}}
+  <img src={{post.imageUrl}} title={{post.imageTitle}}>
+{{/if}}
+
+{{post.text}}
+```
+```js
+// components/post.js
+export default Component.extend({
+  tagName: 'section',
+  classNames: ['post'],
+  classNameBindings: ['type'],
+  ariaRole: 'region',
+
+  /* Arguments */
+  post: null,
+
+  type: readOnly('post.type'),
+
+  didInsertElement() {
+    this._super(...arguments);
+    if (this.type === 'image') {
+      setupImageOverlay(this.element.querySelector('img'));
+    }
+  }
+});
+```
+
+And here is an equivalent component written with the Glimmer component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+<section ...attributes role="region" type={{@post.type}} class="post {{@post.type}}">
+  {{#if (eq @post.type 'image')}}
+    <img
+      {{did-render (action this.didInsertImage)}}
+      src={{@post.imageUrl}}
+      title={{@post.imageTitle}}
+    />
+  {{/if}}
+
+  {{@post.text}}
+</section>
+```
+```js
+// components/post.js
+export default class PostComponent extends GlimmerComponent {
```

Comment:
This might be outside the scope of the RFC, but in (and post-)Octane, do you expect `GlimmerComponent` to be the commonly-used identifier for this base class? Of course users can import it however they like, but I imagine the guides and blueprints will all standardize on one particular name.

In the near term I can see the value in distinguishing between Glimmer components and classic Ember ones (though in many codebases I imagine the ES class syntax will also be a differentiator), but in the long term do you see `GlimmerComponent` as the standard name for these things?

Reply:
I’d personally think the common thing would be to `import Component from ‘special-place’` (and not have each import duplicate the base class type/name), and rely on the import path to be the signifier. In most modules I would expect the import to be pretty close to the class extend line so I don’t really see this as an issue...

Reply:
I agree, and I think it's telling that `sparkles-component` users (at least in the cases I've checked) write `import Component` rather than `import SparklesComponent`.

Reply:
I’ll update the examples accordingly, including imports so folks see the whole picture

Reply:
From a learning and context switching perspective it may make sense to have the blueprint (and taught) import class name be `GlimmerComponent` this makes it easier to distinguish and understand the differences for existing users.

The flip side of that is that there's some legacy baggage for new users or Greenfield projects that only use Glimmer style components.

Reply:
One distinguishing factor will be that classic components will _probably_ still be defined using classic class syntax in the official guides. Even if the decorators RFC is merged, it does not include legacy decorators (as discussed in the RFC) and they will only be available via addons like `ember-decorators`, so the official guides will not be able to use them (or we would have to make the addons official).